### PR TITLE
Ckan 2.9 uploads: validate mimetypes of files uploaded from disk, including zip files

### PR DESCRIPTION
**Pull request by KatiRG**
## What this PR accomplishes
Determines the mimetype of files uploaded from disk, including for files contained within a `.zip` file, using [python-magic](https://github.com/ahupp/python-magic), an interface to the libmagic file type identification library that identifies file types by checking their headers.

### File types tested:
- .txt, .csv, .docx, .xlsx, .exe, .kml
- .exe file with .exe extension removed
- .zip file containing accepted file types
- .zip file where one of the files is an exe file

The script logic is as follows:

### For file uploaded from disk
- As in the original script, if the filename includes an extension, this extension is checked against an [allowed list of extensions/filetypes](https://github.com/ongov/ckanext-ontario_theme/blob/ckan_2.9_uploads/ckanext/ontario_theme/accepted_resource_formats.json) and rejected if not found in that list
- If the extension is valid, or if the filename has no extension, the mimetype is determined using python-magic, then that mimetype is checked against the same list of allowed filetypes and rejected if not in the list
- If the extension is `.zip`, the zip file is opened and, for each file in the zip, the mimetype is determined with python-magic and checked against the allowed list

### For files uploaded via URL
No changes were made for this case. The original logic is as follows:
- the mimetype of the file in the URL is checked using [ckan mimetype_guess](https://docs.ckan.org/en/2.9/maintaining/configuration.html#ckan-mimetype-guess). According to the ckan docs, if the output of this guess is `file_contents`, then the guess can be inaccurate.
- [If the guessed type is 'file_ext'](https://github.com/ongov/ckanext-ontario_theme/blob/main/ckanext/ontario_theme/resource_upload.py#L78), then python [mimetypes](https://docs.python.org/3/library/mimetypes.html) is used to guess the mimetype and this gets assigned to `self.mimetype`. 
- if the file in the URL is a zip, nothing is checked

## Issue addressed
#230 

## What needs review
- the original list of allowed extensions has been adapted in [accepted_resource_formats()](https://github.com/ongov/ckanext-ontario_theme/blob/ckan_2.9_uploads/ckanext/ontario_theme/resource_upload.py#L37) to return both an array of allowed extensions and an array of allowed file types, to avoid code duplication
- the [allowed list of extensions/filetypes](https://github.com/ongov/ckanext-ontario_theme/blob/ckan_2.9_uploads/ckanext/ontario_theme/accepted_resource_formats.json) has been expanded to include ESRI files
- there is a [known bug](https://bugs.astron.com/view.php?id=174) in the `libmagic` library (underneath python-magic) where  CSV files are identified as "application/csv" instead of registered "text/csv", and also for type [application/x-xz](https://bugzilla.redhat.com/show_bug.cgi?id=1947317). I have not tested on the different versions of Ubuntu we are running ([this post suggests checking the output of file itself](https://bytemeta.vip/repo/ahupp/python-magic/issues/263)). For now, I made a [dirty hack in allowed_filetypes()](https://github.com/ongov/ckanext-ontario_theme/blob/ckan_2.9_uploads/ckanext/ontario_theme/resource_upload.py#L63).
- XML files are identified as `text/xml` and not as `application/xml` in the accepted list, so any `text/xml` is passed as part of the hack
- if the mimetype is not identifiable, it is set as None (original code). Should None mimetypes be allowed?
- to pass the test [test_resource_upload.py](https://github.com/ongov/ckanext-ontario_theme/blob/ckan_2.9_uploads/ckanext/ontario_theme/tests/test_resource_upload.py#L107), the mimetype had to be [passed in the args dict itself](https://github.com/ongov/ckanext-ontario_theme/commit/e5e2f11c0cbf85590d59cb253e774488ad94c43b#diff-ebcba8fb5bad7961874c4f37a11514394dfd7eae04bc4d90ef8e7d57c1f22eaaR107), rather than removing it from the returned result and then asserting it to "application/json" which produces `AssertionError: assert 'text/plain' == 'application/json'`

### CODE REVIEW

**Comment by KatiRG** [comment]
Update: for Ubuntu 18.04, the dirty hack for `text/plain` is not needed: CSVs of this type are correctly recognized as such and not as `application/csv` as in Ubuntu 20.04.

**Review by aimalkhan — APPROVED** [review]
Looks good. needs a little clean-up in the comments section. might need a bit of refactoring to make the init method shorter.

File: ckanext/ontario_theme/resource_upload.py [hunk 1]
```diff
@@ -99,35 +127,42 @@ def __init__(self, resource):
...
-                    (self.filename))
-                # remove file - by default a resource can be added without any
-                # values
-                resource['url'] = None
-                resource['url_type'] = ''
-                raise logic.ValidationError(
-                    {'upload':
-                     ['Invalid upload file format, file has been removed.']}
-                )
+            # MODIFICATION START: check extension
```
[the remark this answers is not in the document]
Comment (KatiRG) [reply ckanext/ontario_theme/resource_upload.py:130]: This was the convention that was there before.

File: ckanext/ontario_theme/resource_upload.py [hunk 2]
```diff
@@ -99,35 +127,42 @@ def __init__(self, resource):
...
+                    fileobj = io.BytesIO(self.upload_file.read())
+
+                    with ZipFile(fileobj) as this_zip:
+                        zip_list = this_zip.namelist()
+                        for zip_item in zip_list:
+                            with this_zip.open(zip_item) as each_file:
+                                try: 
+                                    each_mimetype = magic.from_buffer(each_file.read(),
+                                                    mime=True)
+                                except:
```
[the remark this answers is not in the document]
Comment (KatiRG) [reply ckanext/ontario_theme/resource_upload.py:154]: Yes. Please go ahead and test it as well.

File: ckanext/ontario_theme/resource_upload.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (KatiRG) [reply ckanext/ontario_theme/resource_upload.py:130]: The Overkill comment is not mine either. We can remove it but it does have some info.

**Review by aimalkhan — APPROVED** [review]
looks clean and readable! 👍

**Pull request merged**